File: Data Structure/Hash_table/49.group_anagrams.js
// Given an array of strings, group anagrams together.
// Example:
// Input: ["eat", "tea", "tan", "ate", "nat", "bat"],
//     Output:
// [
//     ["ate","eat","tea"],
//     ["nat","tan"],
//     ["bat"]
// ]

/**
 * @param {string[]} strs
 * @return {string[][]}
 */

// Basic Idea， 用排序过的STRING当作KEY
var groupAnagrams = function(strs) {
    if (strs === null) return strs;

    let dic = new Map();
    for (let i = 0; i < strs.length; i++) {
        let arr = strs[i].split('');
        arr.sort();
        let s = arr.join('');
        if (dic.has(s)) {
            let resultArr = dic.get(s);
            resultArr.push(strs[i]);
            dic.set(s, resultArr);
        } else {
            dic.set(s, [strs[i]]);
        }
    }

    let res = []
    for (let value of dic.values()) {
        res.push(value.slice(0, value.length));
    }
    return res;
};

File: Java/49.group_anagrams.java
//Input: ["eat", "tea", "tan", "ate", "nat", "bat"],
//        Output:
//        [
//        ["ate","eat","tea"],
//        ["nat","tan"],
//        ["bat"]
//        ]

// Hashmap/Hashstring method
class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        List<List<String>> res = new ArrayList<>();
        if (strs == null || strs.length == 0) return res;

        HashMap<String, List<String>> dic = new HashMap<>();
        for (int i = 0; i < strs.length; i++) {
            String curWord = strs[i];
            char[] tempArr = curWord.toCharArray();
            Arrays.sort(tempArr);
            String standardWord = new String(tempArr);

            // 这里还是出了Bug，初始化的时候忘记加自己进去
//            if (dic.containsKey(standardWord)) {
//                List<String> tempList = dic.get(standardWord);
//                if (!tempList.contains(strs[i])) {
//                    tempList.add(curWord);
//                }
//                dic.put(standardWord, tempList);
//            } else {
//               dic.put(standardWord, new ArrayList<>());
//            }

            if (!dic.containsKey(standardWord)) dic.put(standardWord, new ArrayList<>());
            dic.get(standardWord).add(curWord);
        }

        for (List<String> value : dic.values()) {
            res.add(new ArrayList(value));
        }
        // dic.forEach((k, v) -> res.add(new ArrayList(v)));

        return res;
    }
}

// Hashstring Count
// Time Complexity： O（NK)
// SPACE COMPLEXITY: O(NK)
class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        List<List<String>> res = new ArrayList<>();
        if (strs == null || strs.length == 0) return res;

        HashMap<String, List<String>> dic = new HashMap<>();

        for (String str : strs) {
            int[] count = new int[26];

            char[] tempArr = str.toCharArray();
            for (char character : tempArr) {
                count[character - 'a']++;
            }

            String hashString = "";
            for (int i = 0; i < count.length; i++) {
                hashString += i;
                for (int j = 0; j < count[i]; j++) {
                    hashString += '#';
                }
            }

            if (!dic.containsKey(hashString)) dic.put(hashString, new ArrayList<>());
            dic.get(hashString).add(str);
        }

        return new ArrayList<List<String>>(dic.values());
    }
}